code/ecdsap256sha2/Hacl.Impl.ECDSA.P256SHA256.Signature.Agile.fst

```diff
@@ -0,0 +1,319 @@
+module Hacl.Impl.ECDSA.P256SHA256.Signature.Agile
+
+open FStar.HyperStack.All
+open FStar.HyperStack
+module ST = FStar.HyperStack.ST
+
+open Lib.IntTypes
+open Lib.Buffer
+open Lib.ByteSequence
+
+open Spec.ECDSA
+open Spec.P256.Lemmas
+
+open FStar.Mul
+open FStar.Math.Lemmas
+
+open Hacl.Hash.SHA2
+
+open Spec.P256
+open Spec.P256.Lemmas
+open Spec.P256.Definitions
+
+open Spec.ECDSAP256.Definition
+
+open Hacl.Impl.LowLevel
+
+open Hacl.Impl.P256
+
+open Hacl.Impl.ECDSA.MM.Exponent
+open Hacl.Impl.ECDSA.MontgomeryMultiplication
+
+open Hacl.Impl.P256.Signature.Common
+
+module H = Spec.Agile.Hash
+module Def = Spec.Hash.Definitions
+
+open Spec.Hash.Definitions
+open Hacl.Hash.Definitions
+
+#set-options "--z3rlimit 100"
+
+inline_for_extraction noextract
+val ecdsa_signature_step12: alg: hash_alg {SHA2_256? alg \/ SHA2_384? alg \/ SHA2_512? alg (* and blake one day *)}
+  ->  mLen: size_t -> m: lbuffer uint8 mLen -> result: felem -> Stack unit
+  (requires fun h -> live h m /\ live h result )
+  (ensures fun h0 _ h1 -> modifies (loc result) h0 h1 /\
+    (
+      assert_norm (pow2 32 < pow2 61);
+      assert_norm (pow2 32 < pow2 125);
+      let hashM = (hashSpec alg) (as_seq h0 m) in 
+      let cutHashM = Lib.Sequence.sub hashM 0 32 in 
+      as_nat h1 result = nat_from_bytes_be cutHashM % prime_p256_order
+    )
+  )
+
+
+let ecdsa_signature_step12 alg mLen m result = 
+  assert_norm (pow2 32 < pow2 61);
+  assert_norm (pow2 32 < pow2 125);
+  push_frame(); 
+    let h0 = ST.get() in 
+  let sz: FStar.UInt32.t = hash_len alg in
+  let mHash = create sz (u8 0) in    
+  
+  begin
```

Comment:
to simplify the reasoning and proof performance, you might benefit from:

```
inline_for_extraction noextract
let statically_agile_hash (a: hash_alg): Hacl.Hash.Definitions.update_st a =
  match a with
  | SHA2_256 -> hash_256 m mLen mHash
  | ...
```
and then you can have your main proof go through with `ifuel 0`, leaving only the definition above to have `ifuel 1`